### PR TITLE
net: coap: Update coap context using payload length

### LINKED ISSUE #40820: coap: blockwise: context current does not match total size after transfer is completed
**Describe the bug**

After downloading a filed using coap block transfer, I noticed that after receiving the last block `ctx->current` does not match ctx->total size.

Example:

File size: 300
Block size: 256

1. Request block 0. 
    block current = 0.
2. Receive block 0, 256 bytes
    block-> current = 256
3. Request block 1.
    block current = 256.
4. Receive block 0, receive 44
    block-> current = 256
5. Transfer completed

**To Reproduce**
Use coap client to receive a file and printout the values of current after the block is received.

**Expected behavior**
Once the whole file is received, the context shall be updated and `current` shall match `total_size`.

**Impact**
Medium

**Logs and console output**
Not available

**Environment (please complete the following information):**
zephyr 1.14
The issue is in the latest version, but it need to be confirmed

### PR DESCRIPTION
**Pull request by efra-mx-aqua**
When a file is download, `ctc->current` is increaed by the block
size, which is correct for all the blocks except the last one.

 Once the download is compelted, ctx->current must match
`ctx->total_size`.

Fixes #40820

### CODE REVIEW

**Review by rlubos — APPROVED** [review]
LGTM

**Review by jukkar — APPROVED** [review]

**Comment by efra-mx-aqua** [comment]
i need to fix the test cases

File: subsys/net/lib/coap/coap.c [hunk 1]
```diff
@@ -1095,6 +1095,10 @@ int coap_next_block_for_option(const struct coap_packet *cpkt,
 	}
 
 	coap_packet_get_payload(cpkt, &block_len);
+	/* Check that the package does not exceed the expected size */
+	if (ctx->total_size < (ctx->current + block_len)) {
```
Comment (rlubos) [inline subsys/net/lib/coap/coap.c:1099]: This isn't right, `total_size` might not be set (i. e. be 0) in case peer didn't send SIZE option.

**Review by rlubos — CHANGES_REQUESTED** [review]

File: subsys/net/lib/coap/coap.c (hunk 1 above)
Comment (efra-mx-aqua) [reply subsys/net/lib/coap/coap.c:1099]: would this work?
```
	if (ctx->total_size > 0 && ctx->total_size < (ctx->current + block_len)) {
		return -EMSGSIZE;
	}
```
Or what do you suggest for that case?

Comment (rlubos) [reply subsys/net/lib/coap/coap.c:1099]: Yes, I think this would be fine.

Comment (efra-mx-aqua) [reply subsys/net/lib/coap/coap.c]: The change was is done

**Comment by efra-mx-aqua** [comment]
I am not able to run the test locally, therefore it difficult to update the test, @rlubos could you help me with that?

**Comment by rlubos** [comment]
> I am not able to run the test locally, therefore it difficult to update the test, @rlubos could you help me with that?

You should be able to build/run a test project just like a regular sample, for instance for `native_posix` you can simply run tests with `west build -t run` locally after building.

As for now, twister reports some build failures, see https://github.com/zephyrproject-rtos/zephyr/runs/4511037614?check_suite_focus=true#step:11:718

**Comment by carlescufi** [comment]
@efra-mx-aqua could you please address the review comments and fix the compliance/tests issues? no need to run the tests locally, you can look at the logs of the CI being run on the server.

**Comment by carlescufi** [comment]
@efra-mx-aqua Would you mind if @rlubos takes your commits and modifies them so that this can be merged?

**Comment by rlubos** [comment]
@efra-mx-aqua Since we're in a bit of urge to close bugs before the release, I took the liberty and submitted an alternative PR with your commits, but with unit tests problems sorted out (#42364), please have a look.

**Comment by rlubos** [comment]
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/42364

**Comment by efra-mx-aqua** [comment]
@efra-mx

**Pull request closed without merging**